src/mono/mono/metadata/marshal.c

```diff
@@ -6580,3 +6580,271 @@ mono_wrapper_caches_free (MonoWrapperCaches *cache)
+	}
+}
+
+static void record_struct_field_physical_lowering (GArray* lowered_bytes, MonoType* type, guint32 offset);
+
+static void record_inlinearray_struct_physical_lowering (GArray* lowered_bytes, MonoClass* klass, guint32 offset) {
+	// Get the first field and record its physical lowering N times
+	MonoClassField* field = mono_class_get_fields_internal (klass, NULL);
+	MonoType* fieldType = field->type;
+	for (int i = 0; i < m_class_inlinearray_value(klass); ++i) {
+		record_struct_field_physical_lowering(lowered_bytes, fieldType, offset + m_field_get_offset(field) + i * mono_type_size(fieldType, NULL));
+	}
+}
+
+static void record_struct_physical_lowering (GArray* lowered_bytes, MonoClass* klass, guint32 offset)
+{
+	if (m_class_is_inlinearray(klass)) {
+		record_inlinearray_struct_physical_lowering(lowered_bytes, klass, offset);
+		return;
+	}
+
+	// Iterate through each field of klass
+	gpointer iter = NULL;
+	MonoClassField* field;
+	while ((field = mono_class_get_fields_internal (klass, &iter))) {
+		if (field->type->attrs & FIELD_ATTRIBUTE_STATIC)
+			continue;
+		if (mono_field_is_deleted (field))
+			continue;
+
+		MonoType* fieldType = field->type;
+		if (fieldType->type == MONO_TYPE_VALUETYPE) {
+			record_struct_physical_lowering(lowered_bytes, mono_class_from_mono_type_internal(fieldType), offset + m_field_get_offset(field));
+		} else {
+			record_struct_field_physical_lowering(lowered_bytes, fieldType, offset + m_field_get_offset(field));
+		}
+	}
+}
+
+static void record_struct_field_physical_lowering (GArray* lowered_bytes, MonoType* type, guint32 offset) {
+	if (type->type == MONO_TYPE_VALUETYPE) {
+		record_struct_physical_lowering(lowered_bytes, mono_class_from_mono_type_internal(type), offset);
+	} else {
+		SwiftPhysicalLoweringKind kind = SWIFT_OPAQUE;
+		if (type->type == MONO_TYPE_I8 || type->type == MONO_TYPE_U8 || type->type == MONO_TYPE_PTR || type->type == MONO_TYPE_FNPTR) {
+			kind = SWIFT_INT64;
+		} else if (type->type == MONO_TYPE_R4) {
+			kind = SWIFT_FLOAT;
+		} else if (type->type == MONO_TYPE_R8) {
+			kind = SWIFT_DOUBLE;
+		}
+
+		set_lowering_range(lowered_bytes, offset, mono_type_size(type, NULL), kind);
+	}
+}
+
+SwiftPhysicalLowering
+mono_marshal_get_swift_physical_lowering (MonoType *type, gboolean native_layout)
+{
+	g_assert (!native_layout);
+	SwiftPhysicalLowering lowering = { 0 };
+
+	// Normalize pointer types to IntPtr.
+	// We don't need to care about specific pointer types at this ABI level.
+	if (type->type == MONO_TYPE_PTR || type->type == MONO_TYPE_FNPTR) {
+		type = m_class_get_byval_arg (mono_defaults.int_class);
+	}
+
+	if (type->type != MONO_TYPE_VALUETYPE && !mono_type_is_primitive(type)) {
+		lowering.byReference = FALSE;
+		return lowering;
+	}
+
+	MonoClass *klass = mono_class_from_mono_type_internal (type);
+
+	GArray* lowered_bytes = g_array_sized_new(FALSE, TRUE, sizeof(SwiftPhysicalLoweringKind), m_class_get_instance_size(klass));
+	
```

Comment:
As of now, `4*PointerSize` is the max, but once we have SIMD support, the max goes up drastically (as we could have up to 4 256-byte vectors here).

I'm using a slightly different algorithm in NativeAOT that I could probably use here instead of matching the CoreCLR algorithm that doesn't allocate "struct size" bytes.

Reply:
ah ok.  I'm not sure we need a completely different algorithm (I didn't look at the NativeAOT one yet, so I don't know how much work it would entail), just rule out cases that are "obviously too big" - whatever that will mean even with SIMD.  i'm particularly (perhaps mistakenly) concerned about InlineArray since it's trivial to make something absolutely massive.